src/template/raw/parse.rs

```diff
@@ -0,0 +1,298 @@
+use super::{EscapedByte, Literal, Part, Template, Variable};
+use crate::error::{self, ParseError};
+
+impl Template {
+	/// Parse the template from a source slice starting at the given position.
+	///
+	/// You must pass the entire source slice and an offset,
+	/// so that source positions in errors are correct.
+	pub fn parse(source: &[u8], start: usize) -> Result<Self, ParseError> {
+		let mut parts = Vec::with_capacity(1);
+		let mut finger = start;
+		while finger < source.len() {
+			let next = match memchr::memchr2(b'$', b'\\', &source[finger..]) {
+				Some(x) => finger + x,
+				None => source.len(),
+			};
+
+			// If we found a non-empty string up to the first backslash or dollar,
+			// then we have a piece of literal text.
+			if next != finger {
+				parts.push(Part::Literal(Literal { range: finger..next }));
+			}
+
+			// If we hit the end of the string, we're done.
+			if next == source.len() {
+				break;
+			}
+
+			// We found an escape sequence.
+			if source[next] == b'\\' {
+				let value = unescape_one(source, next)?;
+				parts.push(Part::EscapedByte(EscapedByte { value }));
+				finger = next + 2;
+
+			// We found a variable substitution.
+			} else {
+				let (variable, end) = Variable::parse(source, next)?;
+				finger = end;
+				parts.push(Part::Variable(variable));
+			}
+		}
+
+		Ok(Self { parts })
+	}
+}
+
+impl Variable {
+	/// Parse a variable from the source.
+	///
+	/// The finger must be the position of the dollar sign in the source.
+	///
+	/// Returns the parsed variable and the index of the byte after the variable.
+	fn parse(source: &[u8], finger: usize) -> Result<(Self, usize), ParseError> {
+		if finger == source.len() {
+			return Err(error::MissingVariableName {
+				position: finger,
+				len: 1,
+			}
+			.into());
+		}
+		if source[finger + 1] == b'{' {
+			Self::parse_braced(source, finger)
+		} else {
+			let name_end = match source[finger + 1..]
+				.iter()
+				.position(|&c| !c.is_ascii_alphanumeric() && c != b'_')
+			{
+				Some(0) => {
+					return Err(error::MissingVariableName {
+						position: finger,
+						len: 1,
+					}
+					.into());
+				},
+				Some(x) => finger + 1 + x,
+				None => source.len(),
+			};
+			let variable = Variable {
+				name: finger + 1..name_end,
+				default: None,
+			};
+			Ok((variable, name_end))
+		}
+	}
+
+	/// Parse a braced variable in the form of "${name[:default]} from source at the given position.
```

Comment:
Could you add an example of this to the README?

Reply:
Hmm? That's an internal function. Nobody can call it directly.

Reply:
I just meant this syntax: `${name[:default]}`.

Reply:
Ah it's there already. I just can't read.

Reply:
Ah, right :)